File: Gramaticas/C/Gramatica_1/G_1.c
/*
Objetivo:
    Revisar si la cadena es capicúa (palíndromo) y solo usa 0 y 1.

Entrada:
    Archivo "g1.txt".

Salida:
    "<cadena> acepta" o "<cadena> NO acepta".
*/

#include <stdio.h>
#include <string.h>

int main() {
    FILE *archivo = fopen("G_1.txt", "r");
    char cadena[100];
    
    while (fgets(cadena, 100, archivo)) {
        cadena[strcspn(cadena, "\n")] = 0; // quitar salto
        
        int n = strlen(cadena);
        int es_binaria = 1;
        int es_palindromo = 1;
        
        for (int i=0; i<n; i++) {
            if (cadena[i] != '0' && cadena[i] != '1') {
                es_binaria = 0;
            }
        }
        for (int i=0; i<n/2; i++) {
            if (cadena[i] != cadena[n-1-i]) {
                es_palindromo = 0;
            }
        }
        
        if (es_binaria && es_palindromo) {
            printf("%s acepta\n", cadena);
        } else {
            printf("%s NO acepta\n", cadena);
        }
    }
    fclose(archivo);
    return 0;
}

File: Gramaticas/Python/Gramatica_1/G_1.py
"""
Objetivo:
    Revisar si la cadena es capicúa (palíndromo) formada solo por 0 y 1.

Entrada:
    Archivo "G_1.txt", una cadena por línea.

Salida:
    "<cadena> acepta" o "<cadena> NO acepta".
"""

# Abrimos el archivo
archivo = open("G_1.txt", "r")

for linea in archivo:
    cadena = linea.strip()  # quitamos salto de línea
    
    # bandera binario
    binario = True
    
    for c in cadena:
        if c != '0' and c != '1':
            binario = False
            break
    
    # verificamos que sea palíndromo
    palindromo = (cadena == cadena[::-1])
    
    if binario and palindromo:
        print(cadena, "acepta")
    else:
        print(cadena, "NO acepta")

archivo.close()
